//
// Copyright 2018 The Project Oak Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

syntax = "proto3";

package oak;

message CreateNodeRequest {
  // The compiled code of the Oak Module to instantiate, in WebAssembly binary
  // format.
  // See https://webassembly.org/docs/binary-encoding/ .
  bytes module = 1;
}

message CreateNodeResponse {
  // The unique ID of the newly created Oak Node.
  string node_id = 1;
  // The port on which the Oak Node is listening.
  int32 port = 2;
}

// Untrusted service in charge of creating Oak Nodes on demand.
service Manager {
  // Request the creation of a new Oak Node with the specified configuration.
  //
  // After the Oak Node is created, the client should connect to the returned
  // endpoint via gRPC and perform a direct attestation against the Node itself,
  // to verify that its configuration corresponds to what the client expects.
  rpc CreateNode(CreateNodeRequest) returns (CreateNodeResponse);
}
